core/iwasm/common/wasm_shared_memory.c

```diff
@@ -435,40 +444,48 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
     if (no_wait) {
         return 1;
     }
-    else {
-        bh_list_status ret;
 
-        if (!(wait_node = wasm_runtime_malloc(sizeof(AtomicWaitNode)))) {
-            wasm_runtime_set_exception(module, "failed to create wait node");
-            return -1;
-        }
-        memset(wait_node, 0, sizeof(AtomicWaitNode));
+    if (!(wait_node = wasm_runtime_malloc(sizeof(AtomicWaitNode)))) {
+        wasm_runtime_set_exception(module, "failed to create wait node");
+        return -1;
+    }
+    memset(wait_node, 0, sizeof(AtomicWaitNode));
 
-        if (0 != os_mutex_init(&wait_node->wait_lock)) {
-            wasm_runtime_free(wait_node);
-            return -1;
-        }
+    if (0 != os_mutex_init(&wait_node->wait_lock)) {
+        wasm_runtime_free(wait_node);
+        return -1;
+    }
 
-        if (0 != os_cond_init(&wait_node->wait_cond)) {
-            os_mutex_destroy(&wait_node->wait_lock);
-            wasm_runtime_free(wait_node);
-            return -1;
-        }
+    if (0 != os_cond_init(&wait_node->wait_cond)) {
+        os_mutex_destroy(&wait_node->wait_lock);
+        wasm_runtime_free(wait_node);
+        return -1;
+    }
 
-        wait_node->status = S_WAITING;
-        os_mutex_lock(&wait_info->wait_list_lock);
-        ret = bh_list_insert(wait_info->wait_list, wait_node);
-        os_mutex_unlock(&wait_info->wait_list_lock);
-        bh_assert(ret == BH_LIST_SUCCESS);
-        (void)ret;
+    wait_node->status = S_WAITING;
+
+    /* acquire the wait info, create new one if not exists */
+    wait_info = acquire_wait_info(address, true, wait_node);
+
+    if (!wait_info) {
+        os_mutex_destroy(&wait_node->wait_lock);
+        wasm_runtime_free(wait_node);
+        wasm_runtime_set_exception(module, "failed to acquire wait_info");
+        return -1;
     }
 
     /* condition wait start */
     os_mutex_lock(&wait_node->wait_lock);
 
-    os_cond_reltimedwait(&wait_node->wait_cond, &wait_node->wait_lock,
-                         timeout < 0 ? BHT_WAIT_FOREVER
-                                     : (uint64)timeout / 1000);
+#if WASM_ENABLE_THREAD_MGR != 0
+    if (!wasm_cluster_is_thread_terminated(exec_env)) {
```

Comment:
Concerned about `if` and not `while` here: [spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup)
@wenyongh @eloparco any thought on what `ret` should be considered as appropriate one to exit loop or what codes should be considered a `false` wakeup?

Reply:
You mean the flag being set between the call to `wasm_cluster_is_thread_terminated` and `os_cond_reltimedwait`, right? That would cause the termination to be missed and the `os_cond_reltimedwait` to wait forever in case of `BHT_WAIT_FOREVER`

Reply:
No, the problem you talking about I solved with your idea of condition variable:
thread 1
- lock `wait_node->wait_lock`
- check whether it's been already terminated
- call wait on the condition variable that releases the lock
thread 2
- we lock `wait_node->wait_lock`
- we call notify on condition variable
- we unlock `wait_node->wait_lock`

To check I ran the tests and it never hangs any more, no tsan warnings as well(besides unrelated LOAD-STORE one)

Reply:
Here I'm concerned about `spurious wakeup`, on some systems thread can actually wake up from wait even there was no other thread in this program that notified it. Usually it's prevented by checking exit condition rather in a loop than just with an `if`.
https://en.wikipedia.org/wiki/Spurious_wakeup

Reply:
Then I presume we want to add a `while` with a check on the `wait_node->status`